perceval/utils/density_matrix.py

```diff
@@ -355,14 +356,104 @@ def normalize(self):
         factor = self.mat.trace()
         self.mat = (1/factor)*self.mat
 
-    def sample(self, count: int = 1):
+    def sample(self, count: int = 1) -> BSSamples:
         """
         sample on the density matrix
         """
         self.normalize()
-        output = random.choices(self.reverse_index, list(self.mat.diagonal()), k=count)
+        samples = random.choices(self.reverse_index, list(self.mat.diagonal()), k=count)
+        output = BSSamples()
+        for state in samples:
+            output.append(state)
         return output
 
+    def measure(self, modes: Union[list[int], int], all_results: bool = True):
+        """
+        makes a measure on a list of modes
+        :param modes: a list of integer for the modes you want to measure
+        :param all_results: whether you want a resulting mixed state or a simple sample
+        """
+        self.normalize()
+        if isinstance(modes, int):
+            modes = [modes]
+
+        if all_results:
```

Comment:
I'm not sure to follow here.

The idea of the method is to make a measurement on a subset of modes.
In the first case, I return every possible measurement with their associated probability, plus the density matrix of the remaining quantum state on the remaining modes (which might still have superposition, etc...).
The second case looks more like what would happen in the real life: I do a measure, the wave function collapse. The probabilistic behavior is emulated by the sample method. (Actually there is a problem here because i forgot to return the measured state)

The two protected method that you cite are two ways of constructing the projectors needed for the wave function collapse. In the first case I need all of them, in the second only one. So I'm not sure what you want me to add in the name of these method.

In any ways, adding this feature was an idea of mine but it's probably not that usefull (it's not in the StateVector measure method). So I don't really know if I keep it.

Reply:
Now, I am confused too. Let's take it up on our discussion tomorrow